### ISSUE TITLE
Update to 4.1.7 fails and leaves NC in maintenance mode

### ISSUE THREAD

**Issue by nursoda** [nextcloud/polls#2793]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### What went wrong, what did you observe?

I just updated the apps in my cloud(s) and got:
```
$ occ app:update --all
polls new version available: 4.1.7
An unhandled exception has been thrown:
Error: Call to undefined method OCA\Polls\Db\OptionMapper::getAll() in /PATH/apps/polls/lib/Db/TableManager.php:317
Stack trace:
#0 /PATH/apps/polls/lib/Migration/RepairSteps/UpdateHashes.php(47): OCA\Polls\Db\TableManager->migrateOptionsToHash()
#1 /PATH/lib/private/Repair.php(127): OCA\Polls\Migration\RepairSteps\UpdateHashes->run()
#2 /PATH/lib/private/legacy/OC_App.php(1068): OC\Repair->run()
#3 /PATH/lib/private/legacy/OC_App.php(1005): OC_App::executeRepairSteps()
#4 /PATH/lib/private/Installer.php(200): OC_App::updateApp()
#5 /PATH/core/Command/App/Update.php(107): OC\Installer->updateAppstoreApp()
#6 /PATH/3rdparty/symfony/console/Command/Command.php(255): OC\Core\Command\App\Update->execute()
#7 /PATH/3rdparty/symfony/console/Application.php(1009): Symfony\Component\Console\Command\Command->run()
#8 /PATH/3rdparty/symfony/console/Application.php(273): Symfony\Component\Console\Application->doRunCommand()
#9 /PATH/3rdparty/symfony/console/Application.php(149): Symfony\Component\Console\Application->doRun()
#10 /PATH/lib/private/Console/Application.php(213): Symfony\Component\Console\Application->run()
#11 /PATH/console.php(100): OC\Console\Application->run()
#12 /PATH/occ(11): require_once('...')
#13 {main}
```
After that, the instance(s) was/were left in maintenance mode.
Additional commands bring the instance and Polls 4.1.7 to working state:
```
$ occ maintenance:mode --off
$ occ upgrade
```

### What did you expect, how polls should behave instead?

It should not break, and break the whole instance.

### What steps does it need to replay this bug?

1. occ app:update polls

### Installation method

Installed/updated using occ

### Installation type

Updated from a minor version within same major version (i.e. 4.0.0 to 4.1.1)

### Affected polls version

4.1.7

### Which browser did you use, when experiencing the bug?

- [X] Firefox
- [ ] Chrome
- [ ] Chromium/Chromium based (i.e. Edge)
- [ ] Safari
- [ ] Other/Don't know

### Which System did you use, when experiencing the bug?

- [ ] Windows
- [X] Linux
- [ ] MacOs
- [ ] iOS
- [ ] Android
- [ ] Other/Don't know

### NC version

Nextcloud 25 (Nextcloud Hub 3)

### PHP engine version

PHP 8.1

### Database engine

MariaDB

### Which user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other/Don't know

### Add your nextcloud server log here

```shell
For that app update, nothing was logged in nextcloud.log
```

**Comment by nursoda** [comment]
OK, if it only happens with OCC since one may fix it. But in setups with many instances and automation, this probably is unacceptable.

**Comment by purerstamp** [comment]
Er same error here when updating from the store :shrug: (and same config as mentioned in the former post)

**Comment by lucge** [comment]
Same problem for me when updating poll from the application menu: I got an error and falls in maintenance mode. To leave the maintenance mode, I had to disable polls and reload my apache server.
Same trace: 
![Screenshot 2023-03-04 08 34 31](https://user-images.githubusercontent.com/14848188/222882748-c426a3f1-1747-4221-be98-612e43266648.png)

**Comment by dartcafe** [comment]
> OK, if it only happens with OCC since one may fix it. But in setups with many instances and automation, this probably is unacceptable.

This error will happen only once, until the classes are updated properly. But how can I override the php caching, to make sure new or changed classes are loaded properly. Assuming that caching is is the reason.

**Comment by dartcafe** [comment]
And yes. It is unacceptable in automated environments. But generally it is a problem, that a crashed app installation leaves the site in maintenance mode.

**Comment by baltaner** [comment]
Hi,
I got the same error - after disable polls-app, restart apache, reenable polls, nextcloud comes up and works
BUT I get following error in logs an when I start occ maintainance:repair
     - ERROR: There is no column with name 'vote_option_hash' on table 'oc_polls_votes'.

`	 [no app in context] Error: Doctrine\DBAL\Schema\SchemaException: There is no column with name 'vote_option_hash' on table 'oc_polls_votes'. at 

	  0. /var/www/nextcloud/3rdparty/doctrine/dbal/src/Schema/Table.php line 335
	     Doctrine\DBAL\Schema\SchemaException::columnDoesNotExist()
	  1. /var/www/nextcloud/3rdparty/doctrine/dbal/src/Schema/Table.php line 242
	     Doctrine\DBAL\Schema\Table->_createIndex()
	  2. /var/www/nextcloud/apps/polls/lib/Db/IndexManager.php line 114
	     Doctrine\DBAL\Schema\Table->addUniqueIndex()
	  3. /var/www/nextcloud/apps/polls/lib/Db/IndexManager.php line 73
	     OCA\Polls\Db\IndexManager->createIndex()
	  4. /var/www/nextcloud/apps/polls/lib/Migration/RepairSteps/CreateIndices.php line 51
	     OCA\Polls\Db\IndexManager->createIndices()
	  5. /var/www/nextcloud/lib/private/Repair.php line 127
	     OCA\Polls\Migration\RepairSteps\CreateIndices->run()
	  6. /var/www/nextcloud/core/Command/Maintenance/Repair.php line 120
	     OC\Repair->run()
	  7. /var/www/nextcloud/3rdparty/symfony/console/Command/Command.php line 255
	     OC\Core\Command\Maintenance\Repair->execute()
	  8. /var/www/nextcloud/3rdparty/symfony/console/Application.php line 1009
	     Symfony\Component\Console\Command\Command->run()
	  9. /var/www/nextcloud/3rdparty/symfony/console/Application.php line 273
	     Symfony\Component\Console\Application->doRunCommand()
	 10. /var/www/nextcloud/3rdparty/symfony/console/Application.php line 149
	     Symfony\Component\Console\Application->doRun()
	 11. /var/www/nextcloud/lib/private/Console/Application.php line 213
	     Symfony\Component\Console\Application->run()
	 12. /var/www/nextcloud/console.php line 100
	     OC\Console\Application->run()
	 13. /var/www/nextcloud/occ line 11
	     require_once("/var/www/nextcloud/console.php")
`

**Comment by dartcafe** [comment]
run `occ polls:de:rebuild`

**Comment by baltaner** [comment]
I get following error after that:
In DbalException.php line 71:

  An exception occurred while executing a query: SQLSTATE[01000]: Warning: 1265 Data truncated for column 'description' at row 24


In ExceptionConverter.php line 114:

  An exception occurred while executing a query: SQLSTATE[01000]: Warning: 1265 Data truncated for column 'description' at row 24


In Exception.php line 26:

  SQLSTATE[01000]: Warning: 1265 Data truncated for column 'description' at row 24


In Connection.php line 82:

  SQLSTATE[01000]: Warning: 1265 Data truncated for column 'description' at row 24
`

**Comment by dartcafe** [comment]
Sorry for inconvenience. Can you check the content of the field `description` of record no. 24?

**Comment by dartcafe** [comment]
Additionally you could update your version with the last [4.1.8](https://github.com/nextcloud/polls/releases/tag/v4.1.8)

**Comment by dartcafe** [comment]
> BUT I get following error in logs an when I start occ maintainance:repair

New learning: `occ maintenance:repair` only runs the `post-migration` repair steps. Database rebuild is a `pre-migration` repair step. F**k

**Comment by baltaner** [comment]
I try to install 4.1.8 later
Thanks a lot! I Like your app!

description-field in record 24 is 'NULL':
`"id","type","title","description","owner","created","expire","deleted","access","anonymous","allow_maybe","vote_limit","show_results","admin_access","important","option_limit","allow_comment","hide_booked_up","allow_proposals","use_no","proposals_expire","misc_settings"

47,datePoll,Zusätzliche Termine Fachdidaktik,,cm,1675924058,0,0,private,0,1,0,always,0,0,0,1,1,disallow,1,0,`

**Comment by dartcafe** [comment]
> I Like your app!

❤️ 

This issues seems to be related to https://github.com/nextcloud/polls/issues/2792#issuecomment-1454047545

For some reason the default value of `TEXT` type colunms can't be set to anything but `null` via DBAL.

**Comment by baltaner** [comment]
Hi,
thanks a lot!
Everything seems to be ok now - no problems with update to 4.1.8, no errors in logs

**Comment by dartcafe** [comment]
👍 Thanks for feedback, but the repairs need still some work. 4.1.9 will come later after testing.

**Comment by schewara** [comment]
Hi
Unfortunately we just ran into this issue while updating to 4.1.8, using the Docker image.

Here the full log entry

```
{
  "reqId": "2OM8KdFkzyLqealq9o2Y",
  "level": 3,
  "time": "2023-03-07T09:41:52+00:00",
  "remoteAddr": "1.2.3.4",
  "user": "username",
  "app": "index",
  "method": "GET",
  "url": "/settings/apps/update/polls",
  "message": "Call to undefined method OCA\\Polls\\Db\\OptionMapper::getAll() in file '/var/www/html/custom_apps/polls/lib/Db/TableManager.php' line 317",
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/110.0",
  "version": "25.0.3.2",
  "exception": {
    "Exception": "Exception",
    "Message": "Call to undefined method OCA\\Polls\\Db\\OptionMapper::getAll() in file '/var/www/html/custom_apps/polls/lib/Db/TableManager.php' line 317",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/html/lib/private/AppFramework/App.php",
        "line": 172,
        "function": "dispatch",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->",
        "args": [
          {
            "__class__": "OCA\\Settings\\Controller\\AppSettingsController"
          },
          "updateApp"
        ]
      },
      {
        "file": "/var/www/html/lib/private/Route/Router.php",
        "line": 298,
        "function": "main",
        "class": "OC\\AppFramework\\App",
        "type": "::",
        "args": [
          "OCA\\Settings\\Controller\\AppSettingsController",
          "updateApp",
          {
            "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
          },
          [
            "polls",
            "settings.AppSettings.updateApp"
          ]
        ]
      },
      {
        "file": "/var/www/html/lib/base.php",
        "line": 1047,
        "function": "match",
        "class": "OC\\Route\\Router",
        "type": "->",
        "args": [
          "/settings/apps/update/polls"
        ]
      },
      {
        "file": "/var/www/html/index.php",
        "line": 36,
        "function": "handleRequest",
        "class": "OC",
        "type": "::",
        "args": []
      }
    ],
    "File": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
    "Line": 165,
    "Previous": {
      "Exception": "Error",
      "Message": "Call to undefined method OCA\\Polls\\Db\\OptionMapper::getAll()",
      "Code": 0,
      "Trace": [
        {
          "file": "/var/www/html/custom_apps/polls/lib/Migration/RepairSteps/UpdateHashes.php",
          "line": 47,
          "function": "migrateOptionsToHash",
          "class": "OCA\\Polls\\Db\\TableManager",
          "type": "->",
          "args": []
        },
        {
          "file": "/var/www/html/lib/private/Repair.php",
          "line": 127,
          "function": "run",
          "class": "OCA\\Polls\\Migration\\RepairSteps\\UpdateHashes",
          "type": "->",
          "args": [
            {
              "__class__": "OC\\Repair"
            }
          ]
        },
        {
          "file": "/var/www/html/lib/private/legacy/OC_App.php",
          "line": 1064,
          "function": "run",
          "class": "OC\\Repair",
          "type": "->",
          "args": []
        },
        {
          "file": "/var/www/html/lib/private/legacy/OC_App.php",
          "line": 1001,
          "function": "executeRepairSteps",
          "class": "OC_App",
          "type": "::",
          "args": [
            "*** sensitive parameters replaced ***",
            [
              "OCA\\Polls\\Migration\\RepairSteps\\RemoveObsoleteMigrations",
              "OCA\\Polls\\Migration\\RepairSteps\\RemoveIndices",
              "OCA\\Polls\\Migration\\RepairSteps\\CreateTables",
              "OCA\\Polls\\Migration\\RepairSteps\\DeleteInvalidRecords",
              "OCA\\Polls\\Migration\\RepairSteps\\UpdateHashes"
            ]
          ]
        },
        {
          "file": "/var/www/html/lib/private/Installer.php",
          "line": 200,
          "function": "updateApp",
          "class": "OC_App",
          "type": "::",
          "args": [
            "*** sensitive parameters replaced ***"
          ]
        },
        {
          "file": "/var/www/html/apps/settings/lib/Controller/AppSettingsController.php",
          "line": 535,
          "function": "updateAppstoreApp",
          "class": "OC\\Installer",
          "type": "->",
          "args": [
            "*** sensitive parameters replaced ***"
          ]
        },
        {
          "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
          "line": 225,
          "function": "updateApp",
          "class": "OCA\\Settings\\Controller\\AppSettingsController",
          "type": "->",
          "args": [
            "*** sensitive parameters replaced ***"
          ]
        },
        {
          "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
          "line": 133,
          "function": "executeController",
          "class": "OC\\AppFramework\\Http\\Dispatcher",
          "type": "->",
          "args": [
            {
              "__class__": "OCA\\Settings\\Controller\\AppSettingsController"
            },
            "updateApp"
          ]
        },
        {
          "file": "/var/www/html/lib/private/AppFramework/App.php",
          "line": 172,
          "function": "dispatch",
          "class": "OC\\AppFramework\\Http\\Dispatcher",
          "type": "->",
          "args": [
            {
              "__class__": "OCA\\Settings\\Controller\\AppSettingsController"
            },
            "updateApp"
          ]
        },
        {
          "file": "/var/www/html/lib/private/Route/Router.php",
          "line": 298,
          "function": "main",
          "class": "OC\\AppFramework\\App",
          "type": "::",
          "args": [
            "OCA\\Settings\\Controller\\AppSettingsController",
            "updateApp",
            {
              "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
            },
            [
              "*** sensitive parameters replaced ***",
              "settings.AppSettings.updateApp"
            ]
          ]
        },
        {
          "file": "/var/www/html/lib/base.php",
          "line": 1047,
          "function": "match",
          "class": "OC\\Route\\Router",
          "type": "->",
          "args": [
            "/settings/apps/update/polls"
          ]
        },
        {
          "file": "/var/www/html/index.php",
          "line": 36,
          "function": "handleRequest",
          "class": "OC",
          "type": "::",
          "args": []
        }
      ],
      "File": "/var/www/html/custom_apps/polls/lib/Db/TableManager.php",
      "Line": 317
    },
    "CustomMessage": "--"
  }
}
```

**Comment by dartcafe** [comment]
@schewara This seems to be a caching or autoload error. 
See https://github.com/nextcloud/polls/issues/2793

**Comment by schewara** [comment]
@dartcafe I think so, that's why I posted it in exactly this issue and didn't create a new one :wink: 

I just wanted to add it for clarification, as in https://github.com/nextcloud/polls/issues/2793#issuecomment-1454742447 it was mentioned that there were no problems with upgrading to 4.1.8, which was not the case for us.

**Comment by dartcafe** [comment]
I was confused, because I thought, I was anserwing in another issue. 😆 
So you just reported the same issue as the original poster, right? 
Didy you repeat the update? It always enabled the app the in my tests.

**Comment by purerstamp** [comment]
Everything's fine here. I had disabled polls app when I first posted (first try to 4.1.7). Today I updated to 4.1.8 while polls app was disabled, then enabled it again. Success! I ran `polls:db:rebuild` after that with no issue, migration's fine too.

Cache had been cleared that far I guess.

**Comment by nursoda** [comment]
I just updated all apps on a production instance. Amongst them was polls, which has been updated from 4.1.6 to 4.1.8. That instance is now down in maintenance mode, on a server where I do not have direct access and need the hosting support (on weekend). This is completely unacceptable. The issue is known for more than a week and the 4.1.8 is still marked "released", and not pulled. Sorry for being pissed.

**Comment by nursoda** [comment]
@dartcafe: The issue is known for more than a week and the 4.1.8 is marked "released". Is the issue fixed with 4.1.8? If so, this ticket should be closed. If not, shouldn't 4.1.7 and 4.1.8 be pulled from the app store?

Sorry for the harsh initial wording of my (now edited) previous post.

**Comment by purerstamp** [comment]
@nursoda this is indeed important to be kind with community which is often doing a better job than corporate... for free. An not only for free, also for "libre" ;)
This issue is frustrating but this app is a lovely FREE version of doodle and cie with no sell of your personal data. Free software can be another way of thinking. As it is free, you can easily find useful info about apps and updates before doing such update by seeing if it seems broken. 
So thanks for the edit :)

@dartcafe thanks for this (not perfect) app!

**Comment by nursoda** [comment]
@purerstamp I'm an app maintainer (yet no dev acutally) myself (twofactor_email), so I know "the other side" well, too. Yet if a dev/maintainer "releases" an app (not as beta), it must be fit for production environments. If it isn't yet, it should not be released (as "release"). We're missing some "production ready labelling" here I guess ;)

**Comment by dartcafe** [comment]
Hello everybody. I was away for some days. 

@purerstamp Thanks for your support, but regardless which words @nursoda was using. I have no problem with that, because I understand his situation. 

@nursoda Retrieving the app is no solution, because it would make the further update paths more complicated. And it seems to be no wide spread issue, since there are only a few reports. The help forum has no reports of this issue until now.

There seem to be two problems: 
1. I guess some caching in OPCache did not update the classes or the proper autoload failed after updating the source base and while executing the repair steps, so that the method is reported as unknown. I have no idea how to solve that in the application without writing a special method just for on update.
2. A failed app update leaves the complete site in maintenance mode, which should not happen. Especially when the app updated is initiated from the appstore. 

I am maintaining a couple of sites, which all updated to 4.1.7 via appstore without any problem. Only updates via scripts using occ ran into this error. Here I just called the script again and everything was fine.

So again sorry for this inconvenience. If someone has an idea how to avoid it in the future...

**Comment by schewara** [comment]
> I was confused, because I thought, I was answering in another issue. laughing So you just reported the same issue as the original poster, right? Did you repeat the update? It always enabled the app the in my tests.

@dartcafe Yes same issue as the original poster. I brought the instance back online by disabling the app and running an upgrade and a `polls:db:rebuild` afterwards just to be on the safe side.

**Comment by stefan-franz** [comment]
After an update of my managed Nextcloud provider, the new download was worked.

**Comment by dartcafe** [comment]
I will close this issue in favor of https://github.com/nextcloud/server/issues/38158 and #2900.
